modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/parser/java/utils/JavaParserUtils.java

```diff
@@ -0,0 +1,385 @@
+
+import org.mule.metadata.api.model.ArrayType;
+import org.mule.runtime.api.connection.ConnectionProvider;
+import org.mule.runtime.api.meta.MuleVersion;
+import org.mule.runtime.extension.api.annotation.Configurations;
+import org.mule.runtime.extension.api.annotation.execution.OnError;
+import org.mule.runtime.extension.api.annotation.execution.OnSuccess;
+import org.mule.runtime.extension.api.annotation.execution.OnTerminate;
+import org.mule.runtime.extension.api.annotation.param.Config;
+import org.mule.runtime.extension.api.annotation.param.Connection;
+import org.mule.runtime.extension.api.annotation.param.Parameter;
+import org.mule.runtime.extension.api.annotation.param.ParameterGroup;
+import org.mule.runtime.extension.api.annotation.source.OnBackPressure;
+import org.mule.runtime.extension.api.exception.IllegalModelDefinitionException;
+import org.mule.runtime.extension.api.runtime.operation.Result;
+import org.mule.runtime.extension.api.runtime.source.Source;
+import org.mule.runtime.extension.api.runtime.streaming.PagingProvider;
+import org.mule.runtime.module.extension.api.loader.java.type.ExtensionElement;
+import org.mule.runtime.module.extension.api.loader.java.type.ExtensionParameter;
+import org.mule.runtime.module.extension.api.loader.java.type.FieldElement;
+import org.mule.runtime.module.extension.api.loader.java.type.MethodElement;
+import org.mule.runtime.module.extension.api.loader.java.type.OperationContainerElement;
+import org.mule.runtime.module.extension.api.loader.java.type.Type;
+import org.mule.runtime.module.extension.api.loader.java.type.TypeGeneric;
+import org.mule.runtime.module.extension.api.loader.java.type.WithAnnotations;
+import org.mule.sdk.api.annotation.DoNotEnforceMinMuleVersion;
+import org.mule.sdk.api.annotation.Extension;
+import org.mule.sdk.api.annotation.MinMuleVersion;
+
+import org.slf4j.Logger;
+
+import javax.inject.Inject;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static java.lang.String.format;
+import static org.slf4j.LoggerFactory.getLogger;
+
+public class JavaParserUtils {
+
+  private static final Logger LOGGER = getLogger(JavaParserUtils.class);
+  public static final MuleVersion FIRST_MULE_VERSION = new MuleVersion("4.1.1");
+
+  public static MuleVersion calculateExtensionMinMuleVersion(ExtensionElement extension) {
+    MuleVersion calculatedMMV = FIRST_MULE_VERSION;
+    if (extension.isAnnotatedWith(Extension.class)) {
+      calculatedMMV = maxMMV(calculatedMMV, new MuleVersion(Extension.class.getAnnotation(MinMuleVersion.class).value()));
+    }
+    if (!(extension.isAnnotatedWith(Configurations.class)
+        || extension.isAnnotatedWith(org.mule.sdk.api.annotation.Configurations.class))) {
+      for (FieldElement field : extension.getFields()) {
+        calculatedMMV = maxMMV(calculatedMMV, calculateFieldMinMuleVersion(field));
+      }
+      for (MethodElement<?> method : extension.getEnclosingMethods()) {
```

Comment:
No

Reply:
I believe it has to

Reply:
Do we support that case where an Extension extends another Extension? If yes, better call this method recursively with the super class (like the other methods) rather than get the inherited methods.

Reply:
We do support that case, but even if we didn't we always need to  recursively evaluate parent classes. These extensions are written in Java and we need to assume that the author made full use of the language capabilities. Sources, operations, configurations and even the extension classes can have parent classes with common stuff. The cloud connectors team does that for once.